File: store.py
"""On-device memory store: hybrid (keyword + vector) search over saved notes.

Each memory is a dict: {id, category, text, source, embedding}.
Persisted to a plain JSON file so it survives restarts.
"""
import os
import json
import threading

import numpy as np
from rank_bm25 import BM25Okapi

import llm
from config import DATA_DIR, DATA_FILE


def _tok(s):
    return s.lower().split()


class Store:
    def __init__(self):
        self.facts = []
        self._lock = threading.Lock()
        self._load()

    def _load(self):
        os.makedirs(DATA_DIR, exist_ok=True)
        if os.path.exists(DATA_FILE):
            with open(DATA_FILE, "r", encoding="utf-8") as f:
                self.facts = json.load(f)

    def _save(self):
        with open(DATA_FILE, "w", encoding="utf-8") as f:
            json.dump(self.facts, f, ensure_ascii=False, indent=2)

    def add(self, text, category, source=""):
        with self._lock:
            emb = llm.embed(text)
            fid = (max(x["id"] for x in self.facts) + 1) if self.facts else 1
            self.facts.append({"id": fid, "category": category,
                               "text": text, "source": source,
                               "embedding": emb})
            self._save()
            return fid

    def update(self, fid, text):
        with self._lock:
            for x in self.facts:
                if x["id"] == fid:
                    x["text"] = text
                    x["embedding"] = llm.embed(text)
                    self._save()
                    return True
            return False

    def delete(self, fid):
        with self._lock:
            n = len(self.facts)
            self.facts = [x for x in self.facts if x["id"] != fid]
            if len(self.facts) < n:
                self._save()
                return True
            return False

    def categories(self):
        c = {}
        for x in self.facts:
            c[x["category"]] = c.get(x["category"], 0) + 1
        return c

    def facts_in(self, category=None):
        return [x for x in self.facts
                if category is None or x["category"] == category]

    def search(self, query, category=None, k=8, with_score=False):
        """Hybrid retrieval: fuse vector-cosine and BM25 rankings via RRF.

        With with_score=True, also returns the best raw cosine similarity — a
        calibrated signal for "is any saved note actually relevant?" that the
        chat layer uses to decide whether to invoke RAG at all.
        """
        pool = self.facts_in(category) or self.facts
        if not pool:
            return ([], 0.0) if with_score else []

        q = np.array(llm.embed(query), dtype=float)
        qn = np.linalg.norm(q) or 1.0
        cosines = {}
        for x in pool:
            a = np.array(x["embedding"], dtype=float)
            cosines[x["id"]] = float(np.dot(q, a) / (qn * (np.linalg.norm(a) or 1.0)))

        vec_ranked = sorted(pool, key=lambda x: cosines[x["id"]], reverse=True)

        bm = BM25Okapi([_tok(x["text"]) for x in pool])
        scores = bm.get_scores(_tok(query))
        bm_ranked = [pool[i] for i in np.argsort(scores)[::-1]]

        # Reciprocal Rank Fusion
        rr = {}
        for rank, x in enumerate(vec_ranked):
            rr[x["id"]] = rr.get(x["id"], 0.0) + 1.0 / (60 + rank)
        for rank, x in enumerate(bm_ranked):
            rr[x["id"]] = rr.get(x["id"], 0.0) + 1.0 / (60 + rank)

        fused = sorted(pool, key=lambda x: rr.get(x["id"], 0.0), reverse=True)[:k]
        # Copies with the raw cosine attached, so callers can drop distractor
        # notes without these extra keys leaking into the persisted facts.
        out = [dict(x, _cos=cosines[x["id"]]) for x in fused]
        if with_score:
            top = max((h["_cos"] for h in out), default=0.0)
            return out, top
        return out


store = Store()

File: mobile/lib/brain/store.dart
/// On-device memory store: hybrid (keyword + vector) search over saved notes —
/// the port of store.py. Each memory is {id, category, text, source,
/// embedding}, persisted to memories.json so it survives restarts.
library;

import 'dart:convert';
import 'dart:io';
import 'dart:math' as math;

import 'embedder.dart';

class Fact {
  int id;
  String category;
  String text;
  String source;
  List<double> embedding;

  /// Raw cosine attached by search (transient — not persisted), so callers can
  /// drop distractor notes; mirrors the `_cos` key in the Python store.
  double cos = 0;

  Fact(this.id, this.category, this.text, this.source, this.embedding);

  Map<String, dynamic> toJson() => {
        'id': id,
        'category': category,
        'text': text,
        'source': source,
        'embedding': embedding,
      };

  static Fact fromJson(Map<String, dynamic> j) => Fact(
        j['id'] as int,
        j['category'] as String,
        j['text'] as String,
        (j['source'] ?? '') as String,
        [for (final e in (j['embedding'] as List)) (e as num).toDouble()],
      );
}

double _dot(List<double> a, List<double> b) {
  var s = 0.0;
  for (var i = 0; i < a.length && i < b.length; i++) {
    s += a[i] * b[i];
  }
  return s;
}

double _norm(List<double> a) {
  final n = math.sqrt(_dot(a, a));
  return n == 0 ? 1.0 : n;
}

double cosine(List<double> a, List<double> b) =>
    _dot(a, b) / (_norm(a) * _norm(b));

List<String> _tok(String s) =>
    s.toLowerCase().split(RegExp(r'\s+')).where((w) => w.isNotEmpty).toList();

/// BM25Okapi over the pool — same defaults as the rank_bm25 package the
/// prototype uses (k1=1.5, b=0.75).
List<double> _bm25Scores(List<List<String>> docs, List<String> query) {
  const k1 = 1.5, b = 0.75;
  final n = docs.length;
  if (n == 0) return const [];
  final avgdl = docs.fold<int>(0, (s, d) => s + d.length) / n;
  final df = <String, int>{};
  for (final d in docs) {
    for (final w in d.toSet()) {
      df[w] = (df[w] ?? 0) + 1;
    }
  }
  return [
    for (final d in docs)
      query.fold<double>(0, (s, w) {
        final f = d.where((x) => x == w).length;
        if (f == 0) return s;
        final idf = math.log((n - df[w]! + 0.5) / (df[w]! + 0.5) + 1);
        return s +
            idf * (f * (k1 + 1)) / (f + k1 * (1 - b + b * d.length / avgdl));
      }),
  ];
}

class MemoryStore {
  final Embedder embedder;
  final String dataFile;
  final List<Fact> facts = [];

  MemoryStore(this.embedder, this.dataFile);

  Future<void> load() async {
    final f = File(dataFile);
    if (!f.existsSync()) return;
    try {
      final raw = json.decode(await f.readAsString()) as List;
      facts
        ..clear()
        ..addAll(raw.map((e) => Fact.fromJson(e as Map<String, dynamic>)));
    } catch (_) {
      // Corrupt file -> start fresh rather than crash the app.
    }
  }

  Future<void> _save() async {
    final f = File(dataFile);
    await f.parent.create(recursive: true);
    await f.writeAsString(json.encode([for (final x in facts) x.toJson()]));
  }

  Future<int> add(String text, String category, {String source = ''}) async {
    final emb = await embedder.embed(text);
    final id = facts.isEmpty
        ? 1
        : facts.map((x) => x.id).reduce(math.max) + 1;
    facts.add(Fact(id, category, text, source, emb));
    await _save();
    return id;
  }

  Future<bool> update(int id, String text) async {
    for (final x in facts) {
      if (x.id == id) {
        x.text = text;
        x.embedding = await embedder.embed(text);
        await _save();
        return true;
      }
    }
    return false;
  }

  Future<bool> delete(int id) async {
    final before = facts.length;
    facts.removeWhere((x) => x.id == id);
    if (facts.length < before) {
      await _save();
      return true;
    }
    return false;
  }

  Map<String, int> categories() {
    final c = <String, int>{};
    for (final x in facts) {
      c[x.category] = (c[x.category] ?? 0) + 1;
    }
    return c;
  }

  List<Fact> factsIn([String? category]) => [
        for (final x in facts)
          if (category == null || x.category == category) x,
      ];

  /// Hybrid retrieval: fuse vector-cosine and BM25 rankings via Reciprocal
  /// Rank Fusion. Returns the fused top-k with each fact's raw cosine set on
  /// `.cos`; `topScore` (best raw cosine) is the calibrated "is any saved note
  /// actually relevant?" signal the chat layer gates RAG with.
  Future<(List<Fact>, double)> search(String query,
      {String? category, int k = 8}) async {
    var pool = factsIn(category);
    if (pool.isEmpty) pool = List.of(facts);
    if (pool.isEmpty) return (const <Fact>[], 0.0);

    final q = await embedder.embed(query);
    final cosines = {for (final x in pool) x.id: cosine(q, x.embedding)};

    final vecRanked = List.of(pool)
      ..sort((a, b) => cosines[b.id]!.compareTo(cosines[a.id]!));

    final scores = _bm25Scores([for (final x in pool) _tok(x.text)], _tok(query));
    final order = List<int>.generate(pool.length, (i) => i)
      ..sort((a, b) => scores[b].compareTo(scores[a]));
    final bmRanked = [for (final i in order) pool[i]];

    final rr = <int, double>{};
    for (var rank = 0; rank < vecRanked.length; rank++) {
      rr[vecRanked[rank].id] = (rr[vecRanked[rank].id] ?? 0) + 1 / (60 + rank);
    }
    for (var rank = 0; rank < bmRanked.length; rank++) {
      rr[bmRanked[rank].id] = (rr[bmRanked[rank].id] ?? 0) + 1 / (60 + rank);
    }

    final fused = List.of(pool)
      ..sort((a, b) => (rr[b.id] ?? 0).compareTo(rr[a.id] ?? 0));
    final out = fused.take(k).toList();
    for (final x in out) {
      x.cos = cosines[x.id]!;
    }
    final top =
        out.isEmpty ? 0.0 : out.map((x) => x.cos).reduce(math.max);
    return (out, top);
  }
}
